File: lovely-ts/dump/engine/particles.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__New = ____lualib.__TS__New
local __TS__Number = ____lualib.__TS__Number
Particles = __TS__Class()
Particles.name = "Particles"
__TS__ClassExtends(Particles, Moveable)
function Particles.prototype.____constructor(self, X, Y, W, H, config)
    Moveable.prototype.____constructor(
        self,
        X,
        Y,
        W,
        H
    )
    config = config or ({})
    self.fill = config.fill
    self.padding = config.padding or 0
    if config.attach then
        if self.role.major == nil then
            error("bad")
        end
        self:set_alignment({major = config.attach, type = "cm", bond = "Strong"})
        table.insert(self.role.major.children, self)
        self.parent = self.role.major
        self.T.x = self.role.major.T.x + self.padding
        self.T.y = self.role.major.T.y + self.padding
        if self.fill then
            self.T.w = self.role.major.T.w - self.padding
            self.T.h = self.role.major.T.h - self.padding
        end
    end
    self.states.hover.can = false
    self.states.click.can = false
    self.states.collide.can = false
    self.states.drag.can = false
    self.states.release_on.can = false
    self.timer = config.timer or 0.5
    self.timer_type = self.created_on_pause and "REAL" or config.timer_type or "REAL"
    self.last_real_time = G.TIMERS[self.timer_type] - self.timer
    self.last_drawn = 0
    self.lifespan = config.lifespan or 1
    self.fade_alpha = 0
    self.speed = config.speed or 1
    self.max = config.max or 1000000000000000
    self.pulse_max = math.min(20, config.pulse_max or 0)
    self.pulsed = 0
    self.vel_variation = config.vel_variation or 1
    self.particles = {}
    self.scale = config.scale or 1
    self.colours = config.colours or ({G.C.BACKGROUND.D})
    if config.initialize then
        do
            local i = 1
            while i <= 60 do
                self.last_real_time = self.last_real_time - 15 / 60
                self:update(15 / 60)
                self:move(15 / 60)
                i = i + 1
            end
        end
    end
    if getmetatable(self) == Particles then
        table.insert(G.I.MOVEABLE, self)
    end
end
function Particles.prototype.update(self, dt)
    if G.SETTINGS.paused and not self.created_on_pause then
        self.last_real_time = G.TIMERS[self.timer_type]
        return
    end
    local added_this_frame = 0
    while G.TIMERS[self.timer_type] > self.last_real_time + self.timer and (#self.particles < self.max or self.pulsed < self.pulse_max) and added_this_frame < 20 do
        self.last_real_time = self.last_real_time + self.timer
        local new_offset = {
            x = self.fill and (0.5 - math.random()) * self.T.w or 0,
            y = self.fill and (0.5 - math.random()) * self.T.h or 0
        }
        if self.fill and self.T.r < 0.1 and self.T.r > -0.1 then
            local newer_offset = {
                x = math.sin(self.T.r) * new_offset.y + math.cos(self.T.r) * new_offset.x,
                y = math.sin(self.T.r) * new_offset.x + math.cos(self.T.r) * new_offset.y
            }
            new_offset = newer_offset
        end
        local ____table_insert_8 = table.insert
        local ____self_particles_7 = self.particles
        local ____temp_1 = math.random() * 2 * math.pi
        local ____temp_2 = math.random() * 2 * math.pi
        local ____temp_3 = math.random() * 0.5 + 0.1
        local ____temp_4 = self.speed * (self.vel_variation * math.random() + (1 - self.vel_variation)) * 0.7
        local ____temp_5 = 0.2 * (0.5 - math.random())
        local ____G_TIMERS_self_timer_type_6 = G.TIMERS[self.timer_type]
        local ____pseudorandom_element_result__0_0 = pseudorandom_element(_G, self.colours)[0]
        if ____pseudorandom_element_result__0_0 == nil then
            ____pseudorandom_element_result__0_0 = {0 / 0, 0 / 0, 0 / 0, 0 / 0}
        end
        ____table_insert_8(____self_particles_7, {
            draw = false,
            dir = ____temp_1,
            facing = ____temp_2,
            size = ____temp_3,
            age = 0,
            velocity = ____temp_4,
            r_vel = ____temp_5,
            e_prev = 0,
            e_curr = 0,
            scale = 0,
            visible_scale = 0,
            time = ____G_TIMERS_self_timer_type_6,
            colour = ____pseudorandom_element_result__0_0,
            offset = new_offset
        })
        added_this_frame = added_this_frame + 1
        if self.pulsed <= self.pulse_max then
            self.pulsed = self.pulsed + 1
        end
    end
end
function Particles.prototype.move(self, dt)
    if G.SETTINGS.paused and not self.created_on_pause then
        return
    end
    Moveable.prototype.move(self, dt)
    if self.timer_type ~= "REAL" then
        dt = dt * G.SPEEDFACTOR
    end
    do
        local i = #self.particles - 1
        while i >= 0 do
            self.particles[i + 1].draw = true
            self.particles[i + 1].e_vel = self.particles[i + 1].e_vel or dt * self.scale
            self.particles[i + 1].e_prev = self.particles[i + 1].e_curr
            self.particles[i + 1].age = self.particles[i + 1].age + dt
            self.particles[i + 1].e_curr = math.min(
                2 * math.min(self.particles[i + 1].age / self.lifespan * self.scale, self.scale * ((self.lifespan - self.particles[i + 1].age) / self.lifespan)),
                self.scale
            )
            self.particles[i + 1].e_vel = (self.particles[i + 1].e_curr - self.particles[i + 1].e_prev) * self.scale * dt + (1 - self.scale * dt) * (self.particles[i + 1].e_vel or 1)
            self.particles[i + 1].scale = self.particles[i + 1].scale + (self.particles[i + 1].e_vel or 1)
            self.particles[i + 1].scale = math.min(
                2 * math.min(self.particles[i + 1].age / self.lifespan * self.scale, self.scale * ((self.lifespan - self.particles[i + 1].age) / self.lifespan)),
                self.scale
            )
            if self.particles[i + 1].scale < 0 then
                table.remove(self.particles, i)
            else
                self.particles[i + 1].offset.x = self.particles[i + 1].offset.x + self.particles[i + 1].velocity * math.sin(self.particles[i + 1].dir) * dt
                self.particles[i + 1].offset.y = self.particles[i + 1].offset.y + self.particles[i + 1].velocity * math.cos(self.particles[i + 1].dir) * dt
                self.particles[i + 1].facing = self.particles[i + 1].facing + self.particles[i + 1].r_vel * dt
                self.particles[i + 1].velocity = math.max(0, self.particles[i + 1].velocity - self.particles[i + 1].velocity * 0.07 * dt)
            end
            i = i + -1
        end
    end
end
function Particles.prototype.fade(self, delay, to)
    G.E_MANAGER:add_event(__TS__New(GameEvent, {
        trigger = "ease",
        timer = self.timer_type,
        blockable = false,
        blocking = false,
        ref_value = "fade_alpha",
        ref_table = self,
        ease_to = to or 1,
        delay = delay
    }))
end
function Particles.prototype.draw(self, alpha)
    alpha = alpha or 1
    prep_draw(_G, self, 1)
    love.graphics.translate(self.T.w / 2, self.T.h / 2)
    for k, w in pairs(self.particles) do
        local v = w
        if v.draw then
            love.graphics.push()
            love.graphics.setColor(v.colour[1], v.colour[2], v.colour[3], v.colour[4] * alpha * (1 - self.fade_alpha))
            love.graphics.translate(v.offset.x, v.offset.y)
            love.graphics.rotate(v.facing)
            love.graphics.rectangle(
                "fill",
                -v.scale / 2,
                -v.scale / 2,
                v.scale,
                v.scale
            )
            love.graphics.pop()
        end
    end
    love.graphics.pop()
    add_to_drawhash(_G, self)
    self:draw_boundingrect()
end
function Particles.prototype.remove(self)
    if self.role.major then
        for k, v in pairs(self.role.major.children) do
            if v == self and type(k) == "number" then
                table.remove(
                    self.role.major.children,
                    __TS__Number(k)
                )
            end
        end
    end
    remove_all(_G, self.children)
    Moveable.prototype.remove(self)
end

File: lovely-ts/dump/engine/particles.ts
///<reference types="lua-types/jit"/>
///<reference types="love-typescript-definitions"/>
///<reference path="../globals.ts"/>

interface ParticleData {
    r_vel: any;
    facing: any;
    velocity: number;
    offset: Position2D;
    draw: boolean;
    e_vel?: number;
    e_prev: number;
    age: number;
    e_curr: number;
    scale: number;
    dir: number;
    colour: HexArray;
    size: number;
    visible_scale: number;
    time: number;
}

class Particles extends Moveable {
    fill: any;
    padding: any;
    timer: any;
    timer_type: any;
    last_real_time: number;
    last_drawn: number;
    lifespan: any;
    fade_alpha: number;
    speed: any;
    max: any;
    pulse_max: number;
    pulsed: number;
    vel_variation: any;
    particles: ParticleData[];
    scale: any;
    colours: HexArray[];
    constructor(X, Y, W, H, config) {
        super(X,Y,W,H)
        config = config || {};
        this.fill = config.fill;
        this.padding = config.padding || 0;
        if (config.attach) {
            if (this.role.major == undefined) {
                error("bad")
            }
            this.set_alignment({ major: config.attach, type: "cm", bond: "Strong" });
            table.insert(this.role.major.children, this);
            this.parent = this.role.major;
            this.T.x = this.role.major.T.x + this.padding;
            this.T.y = this.role.major.T.y + this.padding;
            if (this.fill) {
                this.T.w = this.role.major.T.w - this.padding;
                this.T.h = this.role.major.T.h - this.padding;
            }
        }
        this.states.hover.can = false;
        this.states.click.can = false;
        this.states.collide.can = false;
        this.states.drag.can = false;
        this.states.release_on.can = false;
        this.timer = config.timer || 0.5;
        this.timer_type = this.created_on_pause && "REAL" || config.timer_type || "REAL";
        this.last_real_time = G.TIMERS[this.timer_type] - this.timer;
        this.last_drawn = 0;
        this.lifespan = config.lifespan || 1;
        this.fade_alpha = 0;
        this.speed = config.speed || 1;
        this.max = config.max || 1000000000000000;
        this.pulse_max = math.min(20, config.pulse_max || 0);
        this.pulsed = 0;
        this.vel_variation = config.vel_variation || 1;
        this.particles = [];
        this.scale = config.scale || 1;
        this.colours = config.colours || [G.C.BACKGROUND.D];
        if (config.initialize) {
            for (let i = 1; i <= 60; i++) {
                this.last_real_time = this.last_real_time - 15 / 60;
                this.update(15 / 60);
                this.move(15 / 60);
            }
        }
        if (getmetatable(this) === Particles) {
            table.insert(G.I.MOVEABLE, this);
        }
    }
    update(dt) {
        if (G.SETTINGS.paused && !this.created_on_pause) {
            this.last_real_time = G.TIMERS[this.timer_type];
            return;
        }
        let added_this_frame = 0;
        while (G.TIMERS[this.timer_type] > this.last_real_time + this.timer && (this.particles.length < this.max || this.pulsed < this.pulse_max) && added_this_frame < 20) {
            this.last_real_time = this.last_real_time + this.timer;
            let new_offset = { x: this.fill && (0.5 - math.random()) * this.T.w || 0, y: this.fill && (0.5 - math.random()) * this.T.h || 0 };
            if (this.fill && this.T.r < 0.1 && this.T.r > -0.1) {
                let newer_offset = { x: math.sin(this.T.r) * new_offset.y + math.cos(this.T.r) * new_offset.x, y: math.sin(this.T.r) * new_offset.x + math.cos(this.T.r) * new_offset.y };
                new_offset = newer_offset;
            }
            table.insert(this.particles, { draw: false, dir: math.random() * 2 * math.pi, facing: math.random() * 2 * math.pi, size: math.random() * 0.5 + 0.1, age: 0, velocity: this.speed * (this.vel_variation * math.random() + (1 - this.vel_variation)) * 0.7, r_vel: 0.2 * (0.5 - math.random()), e_prev: 0, e_curr: 0, scale: 0, visible_scale: 0, time: G.TIMERS[this.timer_type], colour: pseudorandom_element(this.colours)[0]??[NaN,NaN,NaN,NaN], offset: new_offset });
            added_this_frame = added_this_frame + 1;
            if (this.pulsed <= this.pulse_max) {
                this.pulsed = this.pulsed + 1;
            }
        }
    };
    move(dt) {
        if (G.SETTINGS.paused && !this.created_on_pause) {
            return;
        }
        Moveable.prototype.move.call(this, dt);
        if (this.timer_type !== "REAL") {
            dt = dt * G.SPEEDFACTOR;
        }
        for (let i = this.particles.length-1; i >= 0; i += -1) {
            this.particles[i].draw = true;
            this.particles[i].e_vel = this.particles[i].e_vel || dt * this.scale;
            this.particles[i].e_prev = this.particles[i].e_curr;
            this.particles[i].age = this.particles[i].age + dt;
            this.particles[i].e_curr = math.min(2 * math.min(this.particles[i].age / this.lifespan * this.scale, this.scale * ((this.lifespan - this.particles[i].age) / this.lifespan)), this.scale);
            this.particles[i].e_vel = (this.particles[i].e_curr - this.particles[i].e_prev) * this.scale * dt + (1 - this.scale * dt) * (this.particles[i].e_vel??1);
            this.particles[i].scale = this.particles[i].scale + (this.particles[i].e_vel??1);
            this.particles[i].scale = math.min(2 * math.min(this.particles[i].age / this.lifespan * this.scale, this.scale * ((this.lifespan - this.particles[i].age) / this.lifespan)), this.scale);
            if (this.particles[i].scale < 0) {
                table.remove(this.particles, i);
            }
            else {
                this.particles[i].offset.x = this.particles[i].offset.x + this.particles[i].velocity * math.sin(this.particles[i].dir) * dt;
                this.particles[i].offset.y = this.particles[i].offset.y + this.particles[i].velocity * math.cos(this.particles[i].dir) * dt;
                this.particles[i].facing = this.particles[i].facing + this.particles[i].r_vel * dt;
                this.particles[i].velocity = math.max(0, this.particles[i].velocity - this.particles[i].velocity * 0.07 * dt);
            }
        }
    };
    fade(delay, to) {
        G.E_MANAGER.add_event(new GameEvent({ trigger: "ease", timer: this.timer_type, blockable: false, blocking: false, ref_value: "fade_alpha", ref_table: this, ease_to: to || 1, delay: delay }));
    };
    draw(alpha) {
        alpha = alpha || 1;
        prep_draw(this, 1);
        love.graphics.translate(this.T.w / 2, this.T.h / 2);
        for (const [k, w] of pairs(this.particles)) {
            let v = w as ParticleData
            if (v.draw) {
                love.graphics.push();
                love.graphics.setColor(v.colour[0], v.colour[1], v.colour[2], v.colour[3] * alpha * (1 - this.fade_alpha));
                love.graphics.translate(v.offset.x, v.offset.y);
                love.graphics.rotate(v.facing);
                love.graphics.rectangle("fill", -v.scale / 2, -v.scale / 2, v.scale, v.scale);
                love.graphics.pop();
            }
        }
        love.graphics.pop();
        add_to_drawhash(this);
        this.draw_boundingrect();
    };
    remove() {
        if (this.role.major) {
            for (const [k, v] of pairs(this.role.major.children)) {
                if (v === this && type(k) === "number") {
                    table.remove(this.role.major.children, Number(k));
                }
            }
        }
        remove_all(this.children);
        Moveable.prototype.remove.call(this);
    };
}